### PR TITLE
Introduce account class selection

### LINKED ISSUE #144: Update OZ contracts to v0.7.0
https://github.com/OpenZeppelin/cairo-contracts/releases/tag/v0.7.0

### LINKED ISSUE #187: Cairo 1 predeployed account contract
## Feature request
I don't know if this is relevant to #183 or #144, but would it be possible to have both Cairo 0 and Cairo 1 OZ account contracts as predeployed contracts on devnet? 

One of the main use cases of `--account-class` would be to test using **any** cairo 1 account, so maybe we could have a solution that does not necessarily requires to provide path to and maintain said contract.

This cairo 1 account contract could be built in alongside cairo 0 (which, truth be told, would double the number of predeployed accounts and make logs a bit more confusing). 
So alternatively, there could be an ability to toggle between implementation version using some other flag.

### PR DESCRIPTION
**Pull request by FabijanC**
## Usage related changes

- Closes #144
  - Upgrades account (defaulting to cairo0 but cairo1 can be selected)
  - Does not update UDC - not available in [OZ 0.7.0](https://github.com/OpenZeppelin/cairo-contracts/releases/tag/v0.7.0)
  - Does not update ERC20 Mintable - not available in OZ 0.7.0
- Closes #187
    - Users can choose between `--account-class cairo0` and `--account-class cairo1`, defaulting to `cairo 0`
- I will deal with #183 in the near future in a separate PR 

## Development related changes

- Simplify tests
  - Wider use of `BackgroundDevnet::get_first_predeployed_account`
- To test the new functionality, I copied old tests and made minor modifications to adapt
- There are some `.expect()` calls in `cli.rs`, but these should virtually never be called
  - One way of fighting this is the exhaustive test in contract_class_choice.rs
- Chargeable account taking care of its own class (not using the same as all predeployed accounts - this PR sets it to cairo 0)
- One place for improvement that I see is the many instances of Cairo 0 class loading. Perhaps it could be better encapsulated?

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet/src/constants.rs [hunk 1]
```diff
@@ -2,14 +2,20 @@ use std::net::{IpAddr, Ipv4Addr};
 
 use starknet_types::chain_id::ChainId;
 
-pub(crate) const CAIRO_0_ACCOUNT_CONTRACT_PATH: &str = concat!(
+pub const CAIRO_0_ACCOUNT_CONTRACT_PATH: &str = concat!(
     env!("CARGO_MANIFEST_DIR"),
     "/accounts_artifacts/OpenZeppelin/0.5.1/Account.cairo/Account.json"
```
Comment (marioiordanov) [inline crates/starknet/src/constants.rs:7]: This is from one of the first PRs, but why not rename this file to Account.casm

File: crates/starknet/src/constants.rs [hunk 2]
```diff
@@ -2,14 +2,20 @@ use std::net::{IpAddr, Ipv4Addr};
...
 );
-
 pub const CAIRO_0_ACCOUNT_CONTRACT_HASH: &str =
     "0x4d07e40e93398ed3c76981e72dd1fd22557a78ce36c0515f679e27f0bb5bc5f";
 
+pub const CAIRO_1_ACCOUNT_CONTRACT_SIERRA_PATH: &str = concat!(
+    env!("CARGO_MANIFEST_DIR"),
+    "/accounts_artifacts/OpenZeppelin/0.7.0/Account.cairo/Account.sierra.json"
```
Comment (marioiordanov) [inline crates/starknet/src/constants.rs:14]: Why not the file name is only Account.sierra, we know that they are in json representation.

File: crates/starknet/accounts_artifacts/OpenZeppelin/0.7.0/Account.cairo/Account.casm.json
Comment (marioiordanov) [inline crates/starknet/accounts_artifacts/OpenZeppelin/0.7.0/Account.cairo/Account.casm.json:1]: I couldn't find where this file is used

Comment (FabijanC) [reply crates/starknet/accounts_artifacts/OpenZeppelin/0.7.0/Account.cairo/Account.casm.json:1]: It's not used, but I was really thinking it should be used somewhere (at least in some tests), so I left it. But I can also remove it.

File: crates/starknet/src/constants.rs (hunk 2 above)
Comment (FabijanC) [reply crates/starknet/src/constants.rs:14]: I just copied what scarb generated

File: crates/starknet/src/constants.rs (hunk 1 above)
Comment (FabijanC) [reply crates/starknet/src/constants.rs:7]: cairo0 artifacts never used the .casm extension, it was always .json

File: crates/starknet/accounts_artifacts/OpenZeppelin/0.7.0/Account.cairo/Account.casm.json
Comment (FabijanC) [reply crates/starknet/accounts_artifacts/OpenZeppelin/0.7.0/Account.cairo/Account.casm.json:1]: Now that I think about it, I think I need to expand the tests by adding a case which gets the class and asserts it's this casm.

File: crates/starknet/src/constants.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet/src/constants.rs:7]: I remember having a similar discussion although not for the same extensions. https://github.com/0xSpaceShard/starknet-devnet-rs/pull/122#discussion_r1285565407

File: crates/starknet/src/constants.rs [hunk 3]
```diff
@@ -2,14 +2,20 @@ use std::net::{IpAddr, Ipv4Addr};
...
+);
+pub const CAIRO_1_ACCOUNT_CONTRACT_SIERRA_HASH: &str =
```
Comment (marioiordanov) [inline crates/starknet/src/constants.rs:16]: If this sierra hash is derived from the CAIRO_1_ACCOUNT_CONTRACT_SIERRA_PATH, why not add a test that asserts that indeed this is the class hash. Simillar to the this test https://github.com/0xSpaceShard/starknet-devnet-rs/blob/10e19ff066fbe7fcfd0dcb303bd110aee77d8d7a/crates/types/src/rpc/contract_class.rs#L263

Comment (FabijanC) [reply crates/starknet/src/constants.rs:16]: will include that in the tests I'm currently adding.

File: crates/starknet/src/utils.rs [hunk 4]
```diff
@@ -49,17 +49,23 @@ pub(crate) mod test_utils {
...
     use crate::utils::exported_test_utils::dummy_cairo_0_contract_class;
 
     pub fn starknet_config_for_test() -> StarknetConfig {
+        let account_contract_class =
+            Cairo0Json::raw_json_from_path(CAIRO_0_ACCOUNT_CONTRACT_PATH).unwrap();
         StarknetConfig {
             seed: DEVNET_DEFAULT_TEST_SEED,
             total_accounts: DEVNET_DEFAULT_TOTAL_ACCOUNTS,
+            account_contract_class: ContractClass::Cairo0(Cairo0ContractClass::RawJson(
+                account_contract_class.clone(),
```
Comment (marioiordanov) [inline crates/starknet/src/utils.rs:66]: Just a small non-important fix, if you swap, the account_contract_class and account_contract_class_hash properties assignment, you can avoid cloning the contract_class.

File: crates/starknet-server/tests/get_transaction_by_hash.rs [hunk 5]
```diff
@@ -16,14 +16,11 @@ mod get_transaction_by_hash_integration_tests {
...
     };
-    use starknet_rs_signers::{LocalWallet, SigningKey};
     use starknet_types::contract_class::Cairo0Json;
     use starknet_types::felt::Felt;
     use starknet_types::traits::ToHexString;
 
-    use crate::common::constants::{
-        CASM_COMPILED_CLASS_HASH, PREDEPLOYED_ACCOUNT_ADDRESS, PREDEPLOYED_ACCOUNT_PRIVATE_KEY,
-    };
+    use crate::common::constants::CASM_COMPILED_CLASS_HASH;
```
Comment (marioiordanov) [inline crates/starknet-server/tests/get_transaction_by_hash.rs:23]: Its from old PR, but what this class hash represents

File: crates/starknet-server/src/cli.rs [hunk 6]
```diff
@@ -93,6 +99,11 @@ impl Args {
                 None => random_number_generator::generate_u32_random_number(),
             },
             total_accounts: self.accounts_count,
+            account_contract_class: self.account_class.get_class().expect("Invalid account class"),
```
Comment (marioiordanov) [inline crates/starknet-server/src/cli.rs:102]: I know that those expects are not goint to fail, but to_starknet_config method is used only inside main function which returns Result. So error propagation should not be a problem when calling it. Thats why the return type of to_starknet_config could be changed to Result<StarknetConfig, anyhow::Error>

File: crates/starknet-server/tests/get_transaction_by_hash.rs (hunk 5 above)
Comment (FabijanC) [reply crates/starknet-server/tests/get_transaction_by_hash.rs:23]:
```rust
// Example contract - Cairo 1
pub const CAIRO_1_CONTRACT_PATH: &str = "test_data/rpc/contract_cairo_v1/output.json";
pub const CASM_COMPILED_CLASS_HASH: &str =
    "0x63b33a5f2f46b1445d04c06d7832c48c48ad087ce0803b71f2b8d96353716ca";
```
I'm guessing it's the casm hash of the cairo contract whose path is `CAIRO_1_CONTRACT_PATH`

File: crates/starknet/src/constants.rs (hunk 3 above)
Comment (FabijanC) [reply crates/starknet/src/constants.rs:16]: Tests expanded in:
- https://github.com/0xSpaceShard/starknet-devnet-rs/pull/196/commits/9d256d0a29fd113f06ad3ebda841ac0abafbd183
- https://github.com/0xSpaceShard/starknet-devnet-rs/pull/196/commits/eaf22d76b37fc78b9752ee081ae0150716855dcf

File: crates/starknet/src/starknet/starknet_config.rs [hunk 7]
```diff
@@ -27,9 +33,15 @@ pub struct StarknetConfig {
 
 impl Default for StarknetConfig {
     fn default() -> Self {
+        let account_contract_class =
+            Cairo0Json::raw_json_from_path(CAIRO_0_ACCOUNT_CONTRACT_PATH).unwrap();
         Self {
             seed: u32::default(),
             total_accounts: u8::default(),
+            account_contract_class: ContractClass::Cairo0(Cairo0ContractClass::RawJson(
+                account_contract_class.clone(),
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/starknet/starknet_config.rs]: Nice tip. Swapped in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/196/commits/66cdb45687f026939f01c4ccca766e28be31c329

File: crates/starknet/src/utils.rs (hunk 4 above)
Comment (FabijanC) [reply crates/starknet/src/utils.rs]: Nice tip. Swapped in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/196/commits/66cdb45687f026939f01c4ccca766e28be31c329

File: crates/starknet/src/constants.rs (hunk 1 above)
Comment (FabijanC) [reply crates/starknet/src/constants.rs]: Renamed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/196/commits/9795932d96e5df85b13acc40259ea158d5718038

File: crates/starknet/src/constants.rs (hunk 2 above)
Comment (FabijanC) [reply crates/starknet/src/constants.rs]: https://github.com/0xSpaceShard/starknet-devnet-rs/pull/196/commits/8b600906170b168daa372b54f99a5fefde453940

File: crates/starknet-server/src/cli.rs (hunk 6 above)
Comment (FabijanC) [reply crates/starknet-server/src/cli.rs]: Looks much better now: https://github.com/0xSpaceShard/starknet-devnet-rs/pull/196/commits/97bf1f0e5da85478dc506632dd2e703cf09dac20

**Review by marioiordanov — APPROVED** [review]

**Pull request merged**